token_filter.module

```diff
@@ -97,19 +211,50 @@ function _token_filter_get_entity_mapping($value_type = 'token', $value = NULL,
   }
 }
 
-
 /**
  * Filter tip callback for the token input filter.
  */
 function _token_filter_filter_tips($filter, $format, $long = FALSE) {
-  // TODO: pass the 'token_types' argument to theme_token_tree* to include
-  // additional tokens when they are officially supported.
+  $replace = isset($filter->settings['replace']) ? $filter->settings['replace'] : 0;
+  $tokens = isset($filter->settings['tokens']) ? $filter->settings['tokens'] : '';
+  $token_list = implode(', ', array_map('trim', explode("\n", $tokens)));
+  $qualifier = '';
+  $text = t('Browse tokens.');
+  if (!empty($token_list)) {
+    if ($replace) {
+      $qualifier = t(', but only tokens of these forms are replaced: ' . check_plain($token_list));
+    }
+    else {
+      $qualifier = t(', but tokens of these forms are ignored: ' . check_plain($token_list));
+    }
+    $text = t('Browse tokens (not all of these may be available).');
+  }
   if ($long) {
-    $output = t('Global tokens will be replaced with their respective token values (e.g. [site:name] or [current-page:title]). The following is a list of the tokens that are available:');
-    $output .= theme('token_tree', array('click_insert' => FALSE));
+    // The "long" description appears on the page filter/tips.
+    $output = t('Global tokens will be replaced with their respective token values (e.g. [site:name] or [current-page:title])!qualifier.', array('!qualifier' => $qualifier));
+    // Only show the token tree once on a page, even if the token filter is used
+    // in multiple text formats.
+    $token_tree = &backdrop_static(__FUNCTION__);
```

Comment:
A static cache is not really needed here. The function that builds the tree (`token_build_tree()`) already uses a static cache. But it does not harm...

Reply:
The point of the static variable is not for caching, but that the token tree only gets rendered once on the `filter/tips` page, even if multiple formats use the Token Filter filter. If a site has, say, 3 different text formats that all use Token Filter, the user doesn't need to see three copies of the full token tree. (In fact, this usage is sort of an anti-cache: normally you cache so that you can use the thing more than once during a page load, but here we're using the static variable to make sure that it doesn't get used more than once.)